src/jrd/replication/Replicator.h

```diff
@@ -0,0 +1,290 @@
+#include "../../common/os/guid.h"
+#include "../../jrd/status.h"
+
+#include "Protocol.h"
+#include "Manager.h"
+
+namespace Replication
+{
+	class Replicator :
+		public Firebird::AutoIface<Firebird::IReplicatedSessionImpl<Replicator, Firebird::CheckStatusWrapper> >,
+		private Firebird::PermanentStorage
+	{
+		typedef Firebird::Array<Firebird::MetaName> MetadataCache;
+		typedef Firebird::HalfStaticArray<SavNumber, 16> SavepointStack;
+
+		struct BatchBlock
+		{
+			Block header;
+			Firebird::UCharBuffer* buffer;
+			MetadataCache metadata;
+			ULONG lastMetaId;
+			ULONG flushes;
+
+			explicit BatchBlock(MemoryPool& pool)
+				: buffer(NULL), metadata(pool),
+				  lastMetaId(MAX_ULONG), flushes(0)
+			{
+				memset(&header, 0, sizeof(Block));
+			}
+
+			ULONG getSize() const
+			{
+				return (ULONG) buffer->getCount();
+			}
+
+			void putTag(UCHAR tag)
+			{
+				buffer->add(tag);
+			}
+
+			void putInt(SLONG value)
+			{
+				const auto newSize = FB_ALIGN(getSize(), FB_ALIGNMENT);
+				buffer->resize(newSize);
+				const auto ptr = (const UCHAR*) &value;
+				buffer->add(ptr, sizeof(SLONG));
+			}
+
+			void putBigInt(SINT64 value)
+			{
+				const auto newSize = FB_ALIGN(getSize(), FB_ALIGNMENT);
+				buffer->resize(newSize);
+				const auto ptr = (const UCHAR*) &value;
+				buffer->add(ptr, sizeof(SINT64));
+			}
+
+			void putMetaName(const Firebird::MetaName& name)
+			{
+				if (lastMetaId < metadata.getCount() && metadata[lastMetaId] == name)
+				{
+					putInt(lastMetaId);
+					return;
+				}
+
+				FB_SIZE_T pos;
+				if (!metadata.find(name, pos))
+				{
+					pos = metadata.getCount();
+					metadata.add(name);
+				}
+
+				putInt(pos);
+				lastMetaId = (ULONG) pos;
+			}
+
+			void putString(const Firebird::string& str)
+			{
+				const auto length = str.length();
+				putInt(length);
+				buffer->add((const UCHAR*) str.c_str(), length);
+			}
+
+			void putBinary(ULONG length, const UCHAR* data)
+			{
+				putInt(length);
+				buffer->add(data, length);
+			}
+		};
+
+		class Transaction :
+			public Firebird::AutoIface<Firebird::IReplicatedTransactionImpl<Transaction, Firebird::CheckStatusWrapper> >
+		{
+		public:
+			explicit Transaction(Replicator* replicator)
+				: m_replicator(replicator), m_data(replicator->getPool())
+			{}
+
+			BatchBlock& getData()
+			{
+				return m_data;
+			}
+
+			// IDisposable methods
+
+			void dispose()
+			{
+				delete this;
+			}
+
+			// IReplicatedTransaction methods
+
+			FB_BOOLEAN prepare()
+			{
+				return m_replicator->prepareTransaction(this) ? FB_TRUE : FB_FALSE;
+			}
+
+			FB_BOOLEAN commit()
+			{
+				return m_replicator->commitTransaction(this) ? FB_TRUE : FB_FALSE;
+			}
+
+			FB_BOOLEAN rollback()
+			{
+				return m_replicator->rollbackTransaction(this) ? FB_TRUE : FB_FALSE;
+			}
+
+			FB_BOOLEAN startSavepoint()
+			{
+				return m_replicator->startSavepoint(this) ? FB_TRUE : FB_FALSE;
+			}
+
+			FB_BOOLEAN releaseSavepoint()
```

Comment:
Firebird 2.5 uses "relative" comit/rollback of savepoints which caused some bugs. Firebird 4 uses "absolute" rollback to savepoint with given number. Shouldn't this piece of code be changed accordingly?

Reply:
AFAIU, you speak about looper savepoints. But these methods behave as SAVEPOINT, RELEASE SAVEPOINT and ROLLBACK TO SAVEPOINT statements. How they're implemented internally is irrelevant, provided that all changes between "start" and "rollback" are undone.

Reply:
I just see that in ensureSavepoint() you start whole stack of savepoints that is unnecessary, IMHO.

Reply:
Priorly replicated savepoints are not replicated again. The rest of the stack is needed to release/rollback them later. I'm open to any optimization ides here.

Reply:
> The rest of the stack is needed to release/rollback them later.

Nope. It is easy to handle release of "never started savepoints" by full rollback. Here is Avalerion's sequence of calls:
1) Transaction stared savepoint 1 - ignored by replication module because transaction is still read-only
2) select started savepoint 2 - the same
3) select release savepoint 2 - the same
4) update started savepoint 3 - still nothing
5) update changed a record - replication modile at last is initialized, allocating replicator (ReplTransaction in Avalerion) and call updateRecord()
6) update rollback or release savepoint 3 - plugin::rollback/releaseSavepoint() is called and replication plugin scan internal list of savepoints (in asynchronous mode only) looking for number 3. When not found, it assumes that whole work from beginning is done or rolled back and act accordingly.

In your case at point 5 (AFAIU) startSavepoint() in replication module will be called several times which is avoidable. Savepoint is replicated at moment when plugin::releaseSavepoint() has been successfully called so additional marking of it as "replicated" is unnecesary because the flag is immediatelly cleaned up by Savepoint::release() which is called right after that.